### ISSUE TITLE
vlang 4.3.0 fails to build: `v.c: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_alloc'?`

### ISSUE THREAD

**Issue by barracuda156** [vlang/v#20507]
### Describe the bug

It is been a few months I compiler `vlang` last time on macOS ppc, and turns out it is broken now. The last version I built was 2023.15.
4.3.0 fails almost immediately on:
```
Using local tcc
Using local vc
make[1]: Entering directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3'
The executable './thirdparty/tcc/tcc.exe' does not work.
make[1]: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3'
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/cc/opt/local/bin/gcc-mp-13 -Os -arch ppc -std=gnu99 -w -o v1.exe /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c -lm -lpthread -Wl,-headerpad_max_install_names -latomic -L/opt/local/lib -lMacportsLegacySupport -arch ppc
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c: In function '__closure_alloc':
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16418:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_alloc'?
16418 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16528:40: note: in expansion of macro '_CLOSURE_SIZE'
16528 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
```

### Reproduction Steps

`sudo port -v install vlang` on a PowerPC system.

### Expected Behavior

`vlang` should build, as it did earlier.

### Current Behavior

```
--->  Building vlang
Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3" && /usr/bin/make -j6 -w all local=true VFLAGS="-v -cg -showcc -show-c-output -prod -cc /opt/local/bin/gcc-mp-13 -cflags -Os" VC=/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc CC="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/cc/opt/local/bin/gcc-mp-13" CXX="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/cxx/opt/local/bin/g++-mp-13" OBJC="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/objc/opt/local/bin/gcc-mp-13" OBJCXX="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/objcxx/opt/local/bin/g++-mp-13" FC="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/fc/opt/local/bin/gfortran-mp-13" F77="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/f77/opt/local/bin/gfortran-mp-13" F90="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/f90/opt/local/bin/gfortran-mp-13" INSTALL="/usr/bin/install -c" 
make: Entering directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3'
Using local tcc
Using local vc
make[1]: Entering directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3'
The executable './thirdparty/tcc/tcc.exe' does not work.
make[1]: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3'
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/cc/opt/local/bin/gcc-mp-13 -Os -arch ppc -std=gnu99 -w -o v1.exe /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c -lm -lpthread -Wl,-headerpad_max_install_names -latomic -L/opt/local/lib -lMacportsLegacySupport -arch ppc
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c: In function '__closure_alloc':
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16418:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_alloc'?
16418 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16528:40: note: in expansion of macro '_CLOSURE_SIZE'
16528 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16418:51: note: each undeclared identifier is reported only once for each function it appears in
16418 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16528:40: note: in expansion of macro '_CLOSURE_SIZE'
16528 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
In file included from /usr/include/string.h:148,
                 from /opt/local/include/LegacySupport/string.h:25,
                 from /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:805:
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c: In function '__closure_init':
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16566:30: error: '__CLOSURE_GET_DATA_BYTES' undeclared (first use in this function); did you mean '__CLOSURE_GET_DATA'?
16566 |         memcpy(_closure_ptr, __CLOSURE_GET_DATA_BYTES, sizeof(__CLOSURE_GET_DATA_BYTES));
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16418:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_init'?
16418 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16569:25: note: in expansion of macro '_CLOSURE_SIZE'
16569 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c: In function '__closure_create':
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16418:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_init'?
16418 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc/v.c:16581:25: note: in expansion of macro '_CLOSURE_SIZE'
16581 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
make: *** [all] Error 1
make: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3'
Command failed:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3" && /usr/bin/make -j6 -w all local=true VFLAGS="-v -cg -showcc -show-c-output -prod -cc /opt/local/bin/gcc-mp-13 -cflags -Os" VC=/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/v-0.4.3/vc CC="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/cc/opt/local/bin/gcc-mp-13" CXX="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/cxx/opt/local/bin/g++-mp-13" OBJC="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/objc/opt/local/bin/gcc-mp-13" OBJCXX="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/objcxx/opt/local/bin/g++-mp-13" FC="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/fc/opt/local/bin/gfortran-mp-13" F77="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/f77/opt/local/bin/gfortran-mp-13" F90="/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_vlang/vlang/work/compwrap/f90/opt/local/bin/gfortran-mp-13" INSTALL="/usr/bin/install -c" 
Exit code: 2
```

### Additional Information/Context

gcc-13.2.0, Macports build environment

### V version

0.4.3

### Environment details (OS name and version, etc.)

macOS 10.6 / ppc

**Comment by barracuda156** [comment]
So, 0.3.4 apparently is good (at least no closure failure), while 0.3.5 already broken.

**Comment by barracuda156** [comment]
Also broken on FreeBSD: https://www.mail-archive.com/freebsd-pkg-fallout@freebsd.org/msg2259602.html

**Comment by barracuda156** [comment]
@felipensp @spytheman @yuyi98 Could someone please look into this? While specific support for powerpc has not been there, it worked as of 0.3.4, but does not work since 0.3.5 anymore.

**Comment by felipensp** [comment]
We'll take care of it. Thanks for reporting.

**Comment by barracuda156** [comment]
@felipensp Great, thank you!

**Comment by tankf33der** [comment]
@barracuda156 - FYI, now vlang compiles or should on ppc out of box.
Only closures crashes because not ported.

**Comment by barracuda156** [comment]
> FYI, now vlang compiles or should on ppc out of box. Only closures crashes because not ported.

@tankf33der Thank you. Do you mean that closures implementation is not needed for the build to work, it only fails at runtime? But then, is vlang usable or just crashes on launch?

**Comment by tankf33der** [comment]
> > FYI, now vlang compiles or should on ppc out of box. Only closures crashes because not ported.
> 
> [@tankf33der](https://github.com/tankf33der) Thank you. Do you mean that closures implementation is not needed for the build to work, it only fails at runtime? But then, is vlang usable or just crashes on launch?

try and tell me what you see.

p.s. did you receive email from me now ?

**Comment by barracuda156** [comment]
@glaubitz Could you please say if you still have 32-bit powerpc hardware with Debian (or any Linux)?

**Comment by glaubitz** [comment]
Yes, we're still building Debian for 32-bit PowerPC.

**Comment by tankf33der** [comment]
> Yes, we're still building Debian for 32-bit PowerPC.

@glaubitz 

Can you give me access for vlang porting?
No sh*t, only vlang.

**Comment by barracuda156** [comment]
@glaubitz It will make sense to start with Linux, and then once that works, I will test it on Darwin, and if there are any issues due to ABI differences etc., we can hopefully work on that with @tankf33der (I only have machines with Darwin on ppc).

**Comment by glaubitz** [comment]
> > Yes, we're still building Debian for 32-bit PowerPC.
> 
> [@glaubitz](https://github.com/glaubitz)
> 
> Can you give me access for vlang porting? No sh*t, only vlang.

You can register for a free account with the [GCC Compile Farm](https://gcc.gnu.org/wiki/CompileFarm).

After receiving your account, you can access the machine gcc121.cfarm.net which is a fast POWER10 machine running Debian unstable for big-endian 64-bit PowerPC. I have root access on that machine and can install any packages you may need for a 32-bit environment.

There are also various other platforms available such as LoongArch, MIPS, SPARC and so on.

**Comment by tankf33der** [comment]
> > > Yes, we're still building Debian for 32-bit PowerPC.
> > 
> > 
> > [@glaubitz](https://github.com/glaubitz)
> > Can you give me access for vlang porting? No sh*t, only vlang.
> 
> You can register for a free account with the [GCC Compile Farm](https://gcc.gnu.org/wiki/CompileFarm).
> 
> After receiving your account, you can access the machine gcc121.cfarm.net which is a fast POWER10 machine running Debian unstable for big-endian 64-bit PowerPC. I have root access on that machine and can install any packages you may need for a 32-bit environment.
> 
> There are also various other platforms available such as LoongArch, MIPS, SPARC and so on.

Wow, OMG. This is a treasure. I was so missing real fast servers.

**Comment by JalonSolov** [comment]
Better than qemu?  :-)

**Comment by tankf33der** [comment]
> Better than qemu? :-)

vlang ppc64le compilation is 15mins on cheap laptop. :)

**Comment by glaubitz** [comment]
> Wow, OMG. This is a treasure. I was so missing real fast servers.

Glad, it's helpful.

> Better than qemu? :-)

Those machines are all native, there is no emulation involved. You therefore get very fast speeds.

> vlang ppc64le compilation is 15mins on cheap laptop. :)

Then it will probably just take one minute on a real POWER system in the GCC Compile Farm.

**Comment by programmingkidx** [comment]
Is anyone else trying to port V to Mac OS 10.4 PowerPC? I have made a lot of progress so I am thinking this issue can be a place to keep patches to V's source code.

**Comment by barracuda156** [comment]
> Is anyone else trying to port V to Mac OS 10.4 PowerPC? I have made a lot of progress so I am thinking this issue can be a place to keep patches to V's source code.

@programmingkidx I can test on 10.6 on powerpc, if you share a tree or a patch-set.

**Comment by tankf33der** [comment]
this is not ELF and not ELFv2, right?

**Comment by programmingkidx** [comment]
Correct. It is Mach-O.

**Comment by barracuda156** [comment]
Thank you, I will try that once I am back to my hardware (in about two weeks, out of station atm).

**Comment by barracuda156** [comment]
Ok, I can confirm that V compiles now (0.5.2 tested) and can build _some_ dependents. It seems to have troubles on 32-bit though with mmap failing in some cases.

**Comment by barracuda156** [comment]
[vfetch](https://github.com/carlosqsilva/vfetch), after some hacking to fix its build and detection:

<img width="669" height="448" alt="Image" src="https://github.com/user-attachments/assets/01228955-448d-49b0-95de-84691c299c11" />
